src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -24,26 +24,79 @@ def self.compile(compile_options)
         Puppet::Pal.in_environment(compile_options['environment'],
+                                   pre_modulepath: boltlib_path,
                                    envpath: Puppet[:environmentpath],
-                                   facts: compile_options['facts']['values'],
-                                   variables: compile_options['variables']['values']) do |pal|
-          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
-
-          # This compiler has been configured with a node containing
-          # the requested environment, facts, and variables, and is used
-          # to compile a catalog in that context from the supplied AST.
-          pal.with_catalog_compiler do |compiler|
-            # We have to parse the AST inside the compiler block, because it
-            # initializes the necessary type loaders for us.
-            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
-
-            compiler.evaluate(ast)
-            compiler.compile_additions
-            compiler.catalog_data_hash
+                                   facts: compile_options['facts']['values']) do |pal|
+          # No conn info for PE targets can just use empty hash if 
+          # https://github.com/puppetlabs/bolt/pull/1770 is accepted
+          fake_config = {
+            'transport' => 'redacted',
+            'transports' => {
+              'redacted' => 'redacted'
+            }
+          }
+          bolt_inv = Bolt::ApplyInventory.new(fake_config)
+          Puppet.override(bolt_inventory: bolt_inv) do
+            Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+            #TODO: get rid of this debug logging...
+            Puppet.warning("Compile Options")
+            Puppet.warning("#{compile_options}")
+            # This compiler has been configured with a node containing
+            # the requested environment, facts, and variables, and is used
+            # to compile a catalog in that context from the supplied AST.
+            pal.with_catalog_compiler do |compiler|
+              #CODEREVIEW: Do we need to make these configurable in the request?
+              Puppet[:strict] = :warning
+              Puppet[:strict_variables] = false
+              vars = Puppet::Pops::Serialization::FromDataConverter.convert(compile_options['variables']['values'])
+
+              # If fact conflicts with plan var name, fact wins, so delete plan var
+              # CODEREVIEW: logging is strange for this. For vars that are overridden, that would
+              # need to happen in orch logs, for facts it would happen here. Is it ok to just
+              # silently overwrite?
+              compile_options['facts']['values'].keys.each {|fact_name| vars.delete(fact_name)}
+              Puppet.warning("FACTS KEYS #{compile_options['facts']['values'].keys}")
+              pal.send(:add_variables, compiler.send(:topscope), vars)
```

Comment:
What's going on here? Same comment wrt metaprograming around a thing built for this use case.

Reply:
https://github.com/puppetlabs/puppetserver/pull/2329/files#r416044603 I will see if i can get around this.

Reply:
I dont think we can get around this. We need to avoid collisions between top level fact names and variables in plan scope. We need to be in the compiler to parse the variables.

Reply:
*short of changes in Puppet.

Reply:
It looks like this likely should have been updated in puppet, not bolt: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pal/pal_impl.rb#L427 so that we

1. aren't metaprogramming around an existing API (since `variables` is already part of the `in_environment` functionality of PAL)

and 

2. We wouldn't have to update this in multiple places.


I would prefer if we didn't block on this for the impending release unless there's concern around safety of the change right now. I don't mind opening tickets and/or doing work to shore up PAL to fit the growing use cases required from bolt so we can improve maintainability, but if we can merge this now and refactor after the release I think we can provide some real user value in PE with the apply feature.

Reply:
Yeah, I think that's really the crux of my comments. Outside of the `require 'bolt'` and maybe the NodeDefinition vs ExpressionBlock vs Program thing, these seem like changes that should be made in Puppet's PAL implementation (specifically the settings overrides, variable work, and `evaluate_ast_node` call). Is it substantially more work to open a PR against Puppet to upstream these changes? I'd be happy to work with someone on those.

Reply:
I think it's only more complicated because of the stage the agent is at for the release. If we need to merge something to Puppet, let's make sure Night's Watch is looped in.

Reply:
I think we should work together with the @puppetlabs/bolt team and come up with an idea for improving PAL API. My thought for this was that if its already shipping with bolt, it should be OK for the first iteration of apply in pe.

Reply:
For the @puppetlabs/bolt team, since these comments may be a bit all-over-the-place. I have two problems with the approach in this PR.
1) AFAICT, Bolt resource compilation is the primary, quite probably only, consumer of PAL. It looks like currently Bolt has a few hacks around the PAL API to enable what it needs. Since those changes aren't upstreamed into PAL proper those hacks need to be copy-pasted here. I would love to see Puppet's version of PAL allow setting variables like you need to, as well as compiling ast code as you need it, and having setting defaults that you can use by default.

2) We are now loading all of Bolt to, afaict, have the apply_target and apply_result PCore types available. Does this mean that functions like `file::readable` are now available? I have a lot of reservations about loading an entire other product within Puppet Server (where will its logging go? etc). Puppet Server's already kind of a Turducken, and now we've added the Christmas Ham... I would _much_ prefer we separate out the PCore types into a Puppet module that server can load, though I understand that disentangling the PCore types from Bolt may be involved.

I understand neither of these will happen in this iteration, but I'm happy to help with both of those efforts if you believe they're helpful long term.

Reply:
Hey @adreyer (I'm unsure who else to ping here?) - am I way off base with my two concerns above?

Reply:
I'm with Justin, now that we're no longer under such a time crunch, if PAL isn't working for us, we should update _it_ and not hack around it. If other people out in the world _are_ using it, they should realize it's far from the beaten path and experimental, and therefore subject to change.

Also +1 to a module. That's the official extension point for stuff like this, and we should set a good example by using it.

Reply:
@justinstoller fwiw on your 2nd concern, we are not loading all of the types/modules that bolt packages internally, just the ones in this module: https://github.com/puppetlabs/bolt/tree/master/bolt-modules/boltlib

Additionally, when we do load the types in this code, they are not loaded via the `bolt/require` call we make: the types are loaded when we include that linked module at the beginning of the modulepath when calling `pal.in_environment`. This should make it so that _only_ the AST compiler in this new `bolt_types` mode would ever load any of the bolt types/functions

Reply:
Additionally, it looks like Cas has updated the new PR to include changes that only require the specific files from within bolt that the module needs to operate: https://github.com/puppetlabs/puppetserver/pull/2332/files#diff-73104a90aa2b4d7f02119c3f544b30c6R35 rather than the whole project.

To take a step back, personally overall I agree with thoughts towards modularizing the parts of bolt we need in PE, so that we could potentially just load a module instead of adding libpaths etc. but these changes completely block the apply block feature. I support refactoring for maintainability/supportability but I would also like to be able to start testing and exercising apply blocks. It seems to me like making the refactors we are talking about could potentially be rather large, and could potentially take up a significant amount of time.

Reply:
Apologies for splitting these PRs... Looks like Alex replied in the new one. Can we move the discussion there? https://github.com/puppetlabs/puppetserver/pull/2332#issuecomment-621481820